### DISCUSSION TITLE
How to set up a development environment for the 4.0 plugin source?

### DISCUSSION THREAD

**Discussion by difflaender** [gwt-plugins/gwt-eclipse-plugin#465] [Q&A] [answered]
I haven't been able to checkout a buildable version of the plugin source since it was picked up again. The instructions in the readme.md are unfortunately not giving enough details, so I am hoping someone else can point out what I may be missing.

So far, I have tried a fresh copy of this plugin's main branch with a vanilla installation of the "Eclipse for Eclipse Committers" bundle (both in versions 2022-09 and 2023-03, but there appear to be no differences in the errors I get).
I've manually installed the TPD plugin from the upsite linked at https://github.com/eclipse-cbi/targetplatform-dsl and imported the source as an "Existing Maven Project".

After using the plugin to set the "gwt-eclipse-2022-09.tpd" as target platform, I am still left with 61 error markers in Eclipse. A lot of them are repeating, but these are the distinct ones:

- Missing artifact com.gwtplugins.eclipse:ide-target-platform.repo:zip:3.0.0-SNAPSHOT
- Project build error: Unknown packaging: eclipse-feature
- Project build error: Unknown packaging: eclipse-plugin
- Project build error: Unknown packaging: eclipse-repository
- Project build error: Unknown packaging: eclipse-target-definition
- Project build error: Unknown packaging: eclipse-test-plugin
- Project build error: Unresolveable build extension: Plugin org.eclipse.tycho:tycho-maven-plugin:3.0.1 or one of its dependencies could not be resolved: Failed to collect dependencies at org.eclipse.tycho:tycho-maven-plugin:jar:3.0.1 -> org.eclipse.tycho:tycho-core:jar:3.0.1 -> org.eclipse.tycho:p2-maven-plugin:jar:3.0.1 -> org.eclipse.platform:org.eclipse.equinox.p2.publisher:jar:1.7.200 -> org.eclipse.platform:org.eclipse.equinox.p2.metadata.repository:jar:[0.0,)

Any suggestions?

**Comment by niloc132** [comment]
It looks like you are checking out the 3.0.0 build instead of the updated source, which is now marked as 4.0.0? Those errors look familiar for the old version, before the maven project setup was updated.

Can you confirm that this builds from the command line with just `mvn verify`?

**Reply by difflaender** [reply]
Can't check it right now because I don't have an actual command line Maven installed on my current PC. Or would a "Run as Maven verify" in Eclipse be enough of an equivalent?

I would assume that I got the right version, though, since the pom.xml says:

```
  <groupId>com.gwtplugins.eclipse</groupId>
  <artifactId>trunk</artifactId>
  <version>4.0.1-SNAPSHOT</version>
```

I also encountered the same issues when I merged my fork from the upstream repo in an attempt to port our customizations to the new release.

**Reply by difflaender** [reply]
Running the "Maven verify" configuration within Eclipse, I get the following error:

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:3.0.1:package-plugin (default-package-plugin) on project com.gwtplugins.gwt.eclipse.sdkbundle.gwt27: D:\WorkingCopies\gwt-eclipse-plugin-main\plugins\com.gwtplugins.gwt.eclipse.sdkbundle.gwt27\build.properties: bin.includes value(s) [gwt-2.7.0/] do not match any files. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
```

**Reply by niloc132** [reply]
It is quite possible that your own changes, or merging from the old history, could potentially cause a problem like this. Your initial message above said that you did a fresh checkout, and then you got errors related to 3.0.0-SNAPSHOT (while as you noted, a fresh checkout declares 4.0.1-SNAPSHOT as its version), so I'm looking for a little more information to understand this difference.

With a fresh checkout, can you run mvn verify, and share the complete log? Can you also confirm what version of maven and java you are running with?

Please note that our Github Actions build runs this same command`*`, and applies for each commit to the repo, so we are reasonably confident that it should build cleanly in this way.

`*` a small lie, we actually run `mvn package` instead, but only because we don't yet have xvfb running to simulate a monitor attached to the build agent. See https://github.com/gwt-plugins/gwt-eclipse-plugin/blob/main/.github/workflows/build.yml#L43

**Reply by difflaender** [reply]
Okay, I did a new fresh checkout via git clone. (Previously I just downloaded the zip of the main branch)
It actually prompted me to install some tycho plugins this time. No idea why that didn't happen on the first imports.
(In both attempts I didn't actively modify anything about the checked out sources at all. The process of importing the projects into Eclipse just changed the output directory in the `.classpath` files from "target/classes" to "bin".)

Now I get the following error in pretty much every project:
- Conflicting lifecycle mapping metadata (project packaging type="eclipse-feature"): Mapping defined in 'org.eclipse.m2e.pde.connector_2.1.2.20221218-1726 [289]' and 'org.sonatype.tycho.m2e_0.10.0.20220926-1324 [540]'. To enable full functionality, remove the conflicting mapping and run Maven->Update Project Configuration.	pom.xml

And still one more error in the pom.xml of the `verify-feature-completeness` project.
- Missing artifact com.gwtplugins.eclipse:ide-target-platform.repo:zip:3.0.0-SNAPSHOT

In my current setup for this project, I use the version of Maven and Java that come bundled with Eclipse 2023-03:
- JustJ OpenJDK 17.0.6
- Maven 3.8.7

Anyway, here the full log output of running as "Maven verify":
[gwt-eclipse-plugin_maven_verify.log](https://github.com/gwt-plugins/gwt-eclipse-plugin/files/11208850/gwt-eclipse-plugin_maven_verify.log)

**Reply by protoism** [reply, marked as answer]
@difflaender I'll try to answer to your (many) questions

> * Conflicting lifecycle mapping metadata (project packaging type="eclipse-feature"): Mapping defined in 'org.eclipse.m2e.pde.connector_2.1.2.20221218-1726 [289]' and 'org.sonatype.tycho.m2e_0.10.0.20220926-1324 [540]'. To enable full functionality, remove the conflicting mapping and run Maven->Update Project Configuration.	pom.xml

This looks like a conflict between tycho and m2e: they both want to handle testing. I'm really not sure how to fix that... we may find a solution, but I'm afraid that it will restrict the dev environment to a specific eclipse / m2e version.
Workaround: disable maven nature and go with straight Eclipse.

> 
> And still one more error in the pom.xml of the `verify-feature-completeness` project.
> 
> * Missing artifact com.gwtplugins.eclipse:ide-target-platform.repo:zip:3.0.0-SNAPSHOT

This error is maven related, so it should go away when maven integration is related

> 
> In my current setup for this project, I use the version of Maven and Java that come bundled with Eclipse 2023-03:
> 
> * JustJ OpenJDK 17.0.6
> * Maven 3.8.7
> 
> Anyway, here the full log output of running as "Maven verify": [gwt-eclipse-plugin_maven_verify.log](https://github.com/gwt-plugins/gwt-eclipse-plugin/files/11208850/gwt-eclipse-plugin_maven_verify.log)

This one is weird...  I ran maven build in debug mode, and what I see is:

```
 Configuring mojo 'com.googlecode.maven-download-plugin:download-maven-plugin:1.2.1:
   (f) outputDirectory = /home/luca/gwt-eclipse-plugin/plugins/com.gwtplugins.gwt.ec
   (f) outputFileName = gwt-2.7.0.zip
   (f) overwrite = true
   (f) readTimeOut = 0
   (f) retries = 2
   (f) session = org.apache.maven.execution.MavenSession@5a4098d5
   (f) skip = false
   (f) skipCache = false
   (f) unpack = true
   (f) url = http://goo.gl/t7FQSn
 -- end configuration --
 Cache is: /home/luca/.m2/repository/.cache/maven-download-plugin
Got from cache: /home/luca/.m2/repository/.cache/maven-download-plugin/gwt-2.8.2.zip
```

Maven is looking for http://goo.gl/t7FQSn (which redirects to  http://storage.googleapis.com/gwt-releases/gwt-2.7.0.zip) but the plugin uses a cached gwt-2.8.2.zip
Whatever the reason (I'd say some change at google) raising the version of maven-download-plugin maybe fixes the issue, but I'm still checking
